### ISSUE TITLE
Based on liferay-npm-bundler-plugin-inject-angular-dependencies I believe we can support peerDependencies

### ISSUE THREAD

**Issue by izaera** [liferay/liferay-npm-build-tools#77]
The idea is to do the same we are doing in [liferay-npm-bundler-plugin-inject-angular-dependencies](https://github.com/liferay/liferay-npm-build-tools/blob/master/packages/liferay-npm-bundler-plugin-inject-angular-dependencies/src/index.js) but for declared `peerDependencies` instead of using a predefined map.

**Comment by izaera** [comment]
OK, so after playing a lot with npm I've found that:

1. peerDependencies just state that a package wants something to be present and if it isn't npm complains and (in runtime) require may fail.
2. When a package is listed as a peerDependency, that's because it is required in some module inside the package and that require is expected to work.
3. The require calls for peerDependency packages are resolved starting from the context of the npm application (that is, the project). 

So, for example, say that: 

- A peerdepends on B 
- B depends on X version 1, 
- The application depends on X version 2
- A requires X

In this case, A receives X at version 2, not 1. That's because it doesn't matter that A peerdepends on B to resolve X, but X is simply resolved from the root of the project.

**Comment by izaera** [comment]
A corollary for the previous comment is that, as long as a version compatible with a peerDependency is declared in any point of the project's dependency tree, the peerDependency will be satisfied because the needed package will appear as a child of root node_modules.

**Comment by izaera** [comment]
Next conclussion is that peerDependency depends on the whole setup of the project and, thus, it's not predictable in advance. That is: a peerDependency may be satisfied by one version now and, after adding a new dependency on the project, it may change to be satisfied by a different version.

**Comment by izaera** [comment]
So, in the context of the portal, the previous paragraph equates to saying that if a peerDependency is resolved to a version, a new deploy of an npm portlet may change that resolution (because it would be equivalent to adding a new dependency to a project).

That means that we need some mechanism to embed information in npm portlets so that the NPM registry in the portal knows how to resolve peer dependencies.

**Comment by izaera** [comment]
That way, if we deploy a portlet P where package A version 1.0.0 has a peerDependency on X version >=1.0.0 and P provides X version 1.0.0 inside it.

And then, we deploy a portlet Q which provides X version 1.1.0, then package A version 1.0.0 of portlet P should resolve X to version 1.1.0, not 1.0.0.

**Comment by izaera** [comment]
That sounds similar to what we do with dependencies, but the problem with peerDependencies is that we don't know the semver expression needed because peerDependencies can be transitive. 

If we look at the example on [this comment](https://github.com/liferay/liferay-npm-build-tools/issues/77#issuecomment-366187204), A is depending on X through B, but it is receiving the X through the application, not B. So, what semver expression should we use? We could naively think that we should use the semver for X inside the project, but that stops being valid when we deploy, because another project may contain the same version of A and B, but provide a different X through a different semver expression.

**Comment by izaera** [comment]
So, the only possible way to resolve peerDependencies is to do it at deploy time, taking into account the whole setup of the portal (all npm portlets deployed at that moment). But, to do that we also need to know which versions of X are compatible with A. 

Probably, the easiest way to do that is to search all npm portlets containing A and satisfying the X peerDependency and get the greatest version and place it as a "fake" dependency of A in its AMD information.

**Comment by yuchi** [comment]
Please have another look at https://github.com/yuchi/liferay-ng-loader-experiments

There’s another way to look at peerDependencies, one that I’ll try to explain here.

We know that a package doesn’t define statically it’s entire package dependency graph. We can therefore treat a package as an input to a `resolveGraph` function which at runtime can generate a graph if a registry is given too.

```js
type Package = {
  name: string,
  version: SemVerVersion,
  dependencies: { [packageName:string]: SemVerRange }
};

type Registry = Array<Package>;

type PackageResolutions = {
  from: { name: string, version: SemVerVersion },
  to: { name: string, version: SemVerVersion }
};

type Graph = Array<PackageResolutions>;

type resolveGraph = (Package, Registry) => Graph;
```

_(continues…)_

**Comment by izaera** [comment]
@yuchi @jbalsas What do you think? Makes sense? 📚

**Comment by izaera** [comment]
Oops 😂

**Comment by izaera** [comment]
Runtime means loader? The "problem" is that our loader doesn't know anything about packages and versions, it just knows how to connect modules and I would prefer not to make it more complex. So, if I can solve it in the portal and leave the loader alone I would do that.

**Comment by yuchi** [comment]
**Note:** I omitted the difference between `dependencies` `devDependencies` `optionalDependencies` `optionalDevDependencies` because they just change which ones to look at (`dev`) or how to treat failures (`optional`).

When we add `peerDependencies` to the mix things change. In fact they are another list of dependencies but they don’t change how we interpret manifests (_package.json_) or how we treat failures. They change the resolution algorithm.

_(continues…)_

**Comment by yuchi** [comment]
Runtime = loader + portal registry

**Comment by izaera** [comment]
So, your solution (loader + registry) is more complete and I also thought about that. But I think we may compute resolutions in the portal without any need to change the loader. Those resolutions could change after every deploy, but still would be valid in the context of a running portal + its current registry.

**Comment by izaera** [comment]
In fact... I think we are doing the same, but you do it in the loader and I do it in the portal. There's nothing about static module resolution that can only be done in the loader but not in the portal because they have the same information.

The only thing we could do in the loader but not in the portal is context-dependent resolution. In the sense that we could resolve a package X to different versions when requested from A depending on who is requiring A in that context. 

But I wonder if that's a good idea because it would make support much more difficult...

**Comment by yuchi** [comment]
To better solve the problem we need first to split the concept of loading a package in two different phases: **loading the implementation** and **instantiating**. This will become important later.

> **Note:** For sake of mental health I’m expecting every package to export a single module so that `package ≈ module`, this speeds up things a lot in the discussion and doesn’t remove any critical complexity. Also we expect every dependency to be actually `require`ed in the package/module.

To load the implementation we could simply skip `peerDependencies`, expecting any *parent package* to have the peerDependencies required.

But when we reach the instantiation phase the knowledge of peerDeps become critical.

> **Mandatory example:**
>
> Project A works with `react@15` and `react-redux@2`.
> Project B works with `react@16` and `react-redux@2`.
> The `react-redux` package has a peerDependency on `react@*`.
>
> The resolution of `react-redux@2` will correctly give us `react-redux@2.0.31`.
>
> The resolution of `react` **inside `react-redux`** can never be correct if we are instantiating `react-redux` only once.

_(continues…)_

**Comment by yuchi** [comment]
So we need to model this behaviour in our types.

```
type Package = {
  name: string,
  version: SemVerVersion,
  dependencies: { [packageName:string]: SemVerRange },
  peerDependencies?: { [packageName:string]: SemVerRange }
};

type Registry = Array<Package>;

type PackageResolutionEndpoint = {
  name: string,
  qualifiedName: string,
  version: SemVerVersion
};

type PackageResolutions = {
  from: PackageResolutionEndpoint,
  to: PackageResolutionEndpoint
};

type Graph = Array<PackageResolutions>;

type resolveGraph = (Package, Graph, Registry) =>  [PackageResolutionEndpoint, Graph];
```

A qualified name is built by composing knowledge of the previous graph:
- the `name` of this resolved package,
- the `version` of this resolved package,
- a **serialized _SortedSet_** of the **qualified names** of peerDependency of this resolved package as extracted from the parent `Graph`.

We need the last part to be sorted in order for qualified names to be predictable.

_(end)_

**Comment by yuchi** [comment]
> The only thing we could do in the loader but not in the portal is context-dependent resolution. In the sense that we could resolve a package X to different versions when requested from A depending on who is requiring A in that context.

The Loader **must** do this IMHO, of course everything must be already prepared in the registry so it knows exactly what to do. In fact the resolved map of dependencies should have qualified names in it so that a package/module that requires `react-redux@^2.0.10` will have `react-redux@2.0.31#{react@16.0.3}` as a concrete dependency. The Loader will be then able to split by `#` to (down)load the implementation but will store the resulting `module` object indexed by the qualified name.

> But I wonder if that's a good idea because it would make support much more difficult...

The Loader requires a big overhaul in terms of error messages and debugging facilities.
I don’t think that this implementation would make debugging _worse_ than what it already is.

IMHO a lot of nice non-throwing warnings could be enough. For example:

```
The package "react-redux@2.0.31" has a peerDependency on "react@*" but no has been found in the require stack. Ensure your project has one.

Require stack:
- react-redux@2.0.31
- @smc/iot-experience-dashlets@13.5.1
- /js/index.js it.smc.iotexperience.dashboard.web@1.0.3
```

**Comment by yuchi** [comment]
Also if a package has a peerDependency that is not met by the current resolution graph (which is the sum of all ancestor’s **shallow** resolution graphs) we could default to the latest version in the registry which matches. So if I don‘t have a `react` dependency in my project I’d get a qualified `react-redux` which points to the latest `react` it supports.

Also remember that peerDependencies are transitive: if **my project** has a dependency on `react-redux-simplifies` (doesn’t exists, just an example) which depends on `react-redux`, then **I am** in charge of having a concrete dependency on `react`.

**Comment by izaera** [comment]
I see what you mean, but this is not feasible in the current moment I'm afraid. 

Also, if we do that, then we should also change the way standard dependencies are resolved because we have the same problem now. It is subtle, but it has happened already to us with metal:

As we are using semver expressions, you may deploy a project P with say:

A1.0.0 -> B^1.0.0
           -> C1.0.0
B1.0.0 -> C^1.0.0
C1.0.0

And a project Q with:

C1.5.0

In this case, the project P ends up having (when instantiated):

A1.0.0 -> B1.0.0
           -> C1.0.0
B1.0.0 -> C1.5.0

So, as you see, A is using a version of C and B a different one which may lead to errors if the project relies on having just one instance of C (for example because it stores shared information inside C or due to instanceof checks and things like that).

I know this is pushing the model to its limits, but we wanted it to dedupe as much as possible because if we are going to instantiate modules locally then we'd better use webpack.

So, if we resolve dependencies based on semver, why would we do a different thing for peerDependencies in the current moment? 

If our basic assumption is that modules don't have internal state and that, as long as semver expressions are satisfied everything is OK, then we don't need to resolve react to two different versions for react-redux@2 (one for project A, another for B) just the best match for its semver expression.

**Comment by yuchi** [comment]
Your case is not an issue at all. If B1 specified a caret range it explicitly asked for that range. If A1 asked for an exact version there's nothing the Loader can do about it.

That was an issue from the dev side.

**Comment by yuchi** [comment]
Please, do not subvert npm semantics. Would lead to baaaaaaaad situations.

**Comment by izaera** [comment]
Hahahaa. I'm not subverting the semantics, I'm saying that nearly every package out there in the world does that kind of things but it ends up working because people use them in node and webpack and, in the end, they can tweak their node_modules until it works.

In our case there's not way to do that and if it fails after deployment we cannot just say "well, the semver expressions of the package you are using were wrong".

But apart from that, I'm not breaking the semver semantics with my peerDependency resolution proposal either. I'm going to satisfy peer dependencies semver correctly though not in the way you expected. In the same way we do with regular dependencies.

**Comment by izaera** [comment]
In the end, the problem of my example is that even if both resolutions are semver correctly (the one in a single project setup and the one with the loader) they are radically different and that may lead to unexpected results.

Also, in our case, the developer has no control over what gets to resolved to what apart from the dependencies in the packages he directly manages, while using npm you can force updates when range semvers are present in inner packages.

**Comment by izaera** [comment]
What I mean is. Why do you say:

> The resolution of react inside react-redux can never be correct if we are instantiating react-redux only once.

As long as it gets a version compatible with its peerDependency semver expression, how does that differ from our regular dependency resolution? 

Am I missing something? 🤔

**Comment by yuchi** [comment]
Yes: both your package and the React-redux you are requiring must access the same version (and instantiation!) of React, otherwise `instanceof React.Component` will break.

**Comment by izaera** [comment]
I know, and that's exactly the problem you may have with `dependencies` too, as I explained before. It happened to us with Metal and we had to change Metal to make it work again.

I just showed you a simple example with packages A, B and C, but you can get more complex setups that end up failing in the same way because you end up having more than one instance of a package inside resolution trees where only one is expected.

This, in the end, is due to the fact that in webpack/node projects you just can have a single version of -say- react, but in the server you may deploy as many as you want and their inner dependencies may get intertwined because of semver expressions, leading to unexpected resolution graphs.

What I'm saying is that is happens both for dependencies and peerDependencies and that if we "fix" one, we should "fix" the other too.

**Comment by izaera** [comment]
I can try to create an example of the problem with regular deps if you don't believe me 😄 .

**Comment by yuchi** [comment]
> I know, and that's exactly the problem you may have with `dependencies` too, as I explained before. It happened to us with Metal and we had to change Metal to make it work again.

The key difference here is that a rogue semver range can result in “wrong” (actually “unexpected”) graphs.

Currently peerDeps result always and only in wrong graphs.

> I just showed you a simple example with packages A, B and C, but you can get more complex setups that end up failing in the same way because you end up having more than one instance of a package inside resolution trees where only one is expected.

Your example AFAIK should give the same result in the current loader and in a npm/yarn installation on filesystem. As far as I recall the exact range should not force that version on parent packages (and therefore the project), the deduplication algorithm should keep the first it encounters alphabetically (which is determinist but unspecified) and move other versions in packages’ own node_modules.

Oh… oooooh…

Now I get it. Since `A` firstly resolves `C@1.0.0` to `v1.0.0`, later `B`’s `C@^1.0.0` matches `C@1.0.0`.

That’s why you get a different graph from node/npm/yarn to portal.

> I can try to create an example of the problem with regular deps if you don't believe me 😄 .

I’d actually be very interested, but spend no more than 5 minutes :)

**Comment by yuchi** [comment]
Anyway, while I do understand the issue, there’s no way to solve this IMHO.

You either have semver-based deduplication or not.

The only other approach is having per plugin (=OSGi bundle) deduplication based on the actually resolved packages in the `.jar`, and best effort deduplication in the portal. Still not covering peerDeps BTW.

**Comment by izaera** [comment]
That's what I meant.

And, if we decide to go with semver-based deduplication, we can implement peerDependencies with my algorithm that turns peerDependencies into injected fake dependencies. 

The only problem is that, even if everything would be semver correctly, it could break (as it happens now with regular dependencies under some limit circumstances).

Still, until now, this has worked pretty well while satisfying angular peerDependencies with the inject-angular-dependencies plugin, so maybe it's worth doing it even if it is not 100% safe...

The other alternative (to be 100% safe) is to change the approach to OSGi bundle resolution but that would equate to using webpack.

Also, there's an intermediate possibility as you suggest that does OSGi bundle resolution but tries to deduplicate when possible but it would need to be explored.

**Comment by yuchi** [comment]
Wait, let me rephrase it a little bit.

You can’t solve errors caused by exact versions in the source of dependencies. There could always be some difference between a FS based resolution (one-time, locked, executed with npm/yarn) and a runtime/dynamic one (the portal).

You **can** solve peerDependencies using, for example, the algorithm I explained and (although badly) implemented in the linked repo.

Moving to a lock-by-bundle (ala webpack) approach _could_ solve both, but I’m not so sure about this.

- - -

Side note: there could be some kind of lock-by-OSGi-feature, where multiple bundles share the same package graph space.

**Comment by yuchi** [comment]
(And an OSGi feature is more or less similar to an LPKG, where multiple JARs are merged into one)

**Comment by izaera** [comment]
So, what we are going to try is this:

1. Step back to make the bundler work like webpack and make each bundle use its own "private" versions of the packages. This will make version resolution exactly behave like the developer's node_modules folder and will fix the issue explained [here](https://github.com/liferay/liferay-npm-build-tools/issues/77#issuecomment-366213363).
2. Then, we will provide a way to support peerDependencies.
3. And finally, we will provide one or more ways to dedupe shared package graph spaces, whether automagically or by configuration. We'll see 🙏 .

I will also make zip files with projects to reproduce the problems and check that they are fixed during this development.

**Comment by yuchi** [comment]
> Step back to make the bundler work like webpack and make each bundle use its own "private" versions of the packages.

This is a big step back. If we are going in this direction than we can just use webpack with something similar to the [DLL Plugin](https://webpack.js.org/plugins/dll-plugin/). This would make all webpack plugins available and therefore you could just trash the whole bundler (which is sad but also a removed burden).

> Then, we will provide a way to support peerDependencies.

If you go on a per-bundle dependencies, with no deduplication, you can solve this quite trivially by just mimicking the on-FS structure.

> And finally, we will provide one or more ways to dedupe shared package graph spaces, whether automagically or by configuration. We'll see

Very curious about what will be doable or not in this whole process.

**Comment by izaera** [comment]
Yes, it would be a big step back. And if the deduplication is not feasible then we can throw away the whole solution 😱 . 

The problem with that would be that if you have _n_ portlets in a page using Angular, you load, parse, and maintain _n_ copies of Angular in memory which can be a huge issue for many configurations (think of machines with scarce resources).

The main reason why I suspect we can't rely on webpack alone is that the portlet architecture is a _rara avis_ and we diverge a bit from the usual web applications in the sense that we must provide some way for unknown portlets to cooperate, while a simple web application has full control of what's happening in the browser.

I'll have a look at the DllPlugin, though.

**Comment by yuchi** [comment]
Indeed this is a problem only for Composite Applications (which portals are a subset of).

I actually expected he OSGi Alliance to give guidance here but was disappointed to see that all effort was spent on webjars.

IMHO a team up with Electron/Atom people and OSGi could provide a better voice against the power of single-point-of-build apps.
